fedimint-core/src/db/mem_impl.rs

```diff
@@ -176,21 +180,25 @@ impl<'a> IDatabaseTransactionOps for MemTransaction<'a> {
 #[apply(async_trait_maybe_send!)]
 impl<'a> IRawDatabaseTransaction for MemTransaction<'a> {
     async fn commit_tx(self) -> Result<()> {
+        let mut data = self.db.data.lock().await;
+        let mut data_copy = data.clone();
         for op in self.operations {
             match op {
                 DatabaseOperation::Insert(insert_op) => {
-                    self.db
-                        .data
-                        .lock()
-                        .unwrap()
-                        .insert(insert_op.key, insert_op.value);
+                    anyhow::ensure!(
+                        data_copy.insert(insert_op.key, insert_op.value) == insert_op.old_value,
```

Comment:
It just occurred to me that this check is not bullet-proof, let's imaging an entry `x=0` at the beginning, we now run the following operations:

* Start dbtx `A`, where `x=0`
* `A` writes `x=1`
* Start dbtx `B` with `x=0`
* `B` writest `x=2`
* `B` commits successfully, `x=2` now
* Start dbtx `C` with `x=2`
* `C` writes `x=0`
* `C` commits successfully, `x=0` now
* `A` commits, should fail due to a write/write conflict, but succeeds with this code

I don't think this is too problematic for how we use the DB, but it might come to bite us in the ass eventually. I think to do this properly entries need version numbers as I described here https://github.com/elsirion/webimint-rs/issues/31. Thoughts @m1sterc001guy?

Reply:
BTW. I recommend https://www.amazon.com/Database-Internals-deep-dive-distributed-systems/dp/1492040347/ . Very good book to read, since we're already implementing our own databases. :D

Thought even in it they don't dive to deep into implementing MVCC...